File: src/main/java/com/ddd/trainings/proposal/domain/ReviewerPolicyFactory.java
package com.ddd.trainings.proposal.domain;

import com.ddd.trainings.idea.domain.IdeaNameVO;
import com.ddd.trainings.proposal.infra.ReviewerRepo;
import java.util.List;
import java.util.Random;
import org.jmolecules.ddd.annotation.Factory;

@Factory
public class ReviewerPolicyFactory {

  private final ReviewerRepo reviewerRepo;
  private final Random random = new Random();

  public ReviewerPolicyFactory(ReviewerRepo reviewerRepo) {
    this.reviewerRepo = reviewerRepo;
  }

  public ReviewerPolicy byKeywords(IdeaNameVO name) {
    return new ReviewerPolicy() {
      @Override
      public ReviewerId selectReviewer() {
        List<Reviewer> candidates = reviewerRepo.findByKeywordsOrderByCurrentReviewsCount(name.getKeywords());
        return candidates.stream()
            .map(Reviewer::getId)
            .findFirst()
            .orElseThrow(() -> new RuntimeException("No available reviewers"));
      }
    };
  }

  public ReviewerPolicy random() {
    return new ReviewerPolicy() {
      @Override
      public ReviewerId selectReviewer() {
        List<Reviewer> candidates = reviewerRepo.findAll();
        if (candidates.isEmpty()) {
          throw new RuntimeException("No available reviewers");
        }

        int randomIndex = random.nextInt(candidates.size());

        return candidates.get(randomIndex).getId();
      }
    };
  }

}

File: src/Ddd/Trainings/Proposal/Domain/ReviewerPolicyFactory.cs
using Ddd.Trainings.Idea.Domain;
using Ddd.Trainings.Proposal.Domain.Ports;

namespace Ddd.Trainings.Proposal.Domain
{
    public class ReviewerPolicyFactory(IReviewerRepo reviewerRepo)
  {
    public IReviewerPolicy ByKeywords(IdeaNameVO name)
        {
            return new ByKeywordsReviewerPolicy(reviewerRepo, name);
        }

        public IReviewerPolicy Random()
        {
            return new RandomReviewerPolicy(reviewerRepo);
        }

        // TODO: implement new policy and test for it
    }
} 